File: app/src/main/java/com/cyclecare/app/data/repository/SettingsRepositoryImpl.kt
package com.cyclecare.app.data.repository

import com.cyclecare.app.data.local.dao.SettingsDao
import com.cyclecare.app.data.local.entity.SettingsEntity
import com.cyclecare.app.domain.model.AppSettings
import com.cyclecare.app.domain.model.TemperatureUnit
import com.cyclecare.app.domain.model.ThemeMode
import com.cyclecare.app.domain.model.UserProfile
import com.cyclecare.app.domain.repository.SettingsRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.time.LocalTime
import java.time.ZoneId
import javax.inject.Inject

class SettingsRepositoryImpl @Inject constructor(
    private val settingsDao: SettingsDao
) : SettingsRepository {

    override fun getSettings(): Flow<AppSettings> {
        return settingsDao.getSettings().map { entity -> entity?.toDomain() ?: AppSettings() }
    }

    override suspend fun updateSettings(settings: AppSettings) {
        settingsDao.insertSettings(settings.toEntity())
    }

    private fun SettingsEntity.toDomain(): AppSettings {
        val fallbackZone = runCatching { ZoneId.systemDefault().id }.getOrDefault("UTC")
        return AppSettings(
            theme = runCatching { ThemeMode.valueOf(theme) }.getOrDefault(ThemeMode.SYSTEM),
            primaryColor = primaryColor,
            averageCycleLength = averageCycleLength,
            averagePeriodLength = averagePeriodLength,
            lutealPhaseLength = lutealPhaseLength,
            temperatureUnit = runCatching { TemperatureUnit.valueOf(temperatureUnit) }.getOrDefault(TemperatureUnit.CELSIUS),
            dateFormat = dateFormat,
            language = language,
            isPinEnabled = isPinEnabled,
            pinHash = pinHash,
            isBiometricEnabled = isBiometricEnabled,
            isPrivacyModeEnabled = isPrivacyModeEnabled,
            hideNotificationContent = hideNotificationContent,
            notificationsEnabled = notificationsEnabled,
            quietHoursEnabled = quietHoursEnabled,
            quietHoursStart = runCatching { LocalTime.parse(quietHoursStart) }.getOrDefault(LocalTime.of(22, 0)),
            quietHoursEnd = runCatching { LocalTime.parse(quietHoursEnd) }.getOrDefault(LocalTime.of(7, 0)),
            defaultReminderTime = runCatching { LocalTime.parse(defaultReminderTime) }.getOrDefault(LocalTime.of(20, 0)),
            medicationReminderTime = runCatching { LocalTime.parse(medicationReminderTime) }.getOrDefault(LocalTime.of(9, 0)),
            hydrationReminderTime = runCatching { LocalTime.parse(hydrationReminderTime) }.getOrDefault(LocalTime.of(12, 0)),
            bodyMetricsReminderTime = runCatching { LocalTime.parse(bodyMetricsReminderTime) }.getOrDefault(LocalTime.of(7, 30)),
            onboardingCompleted = onboardingCompleted,
            pregnancyMode = pregnancyMode,
            breastfeedingMode = breastfeedingMode,
            menopauseMode = menopauseMode,
            userProfile = UserProfile(
                name = profileName,
                birthYear = profileBirthYear,
                cycleLengthHint = averageCycleLength,
                periodLengthHint = averagePeriodLength,
                tryingToConceive = profileTryingToConceive,
                timezoneId = profileTimezoneId.ifBlank { fallbackZone }
            )
        )
    }

    private fun AppSettings.toEntity(): SettingsEntity {
        return SettingsEntity(
            id = 1,
            theme = theme.name,
            primaryColor = primaryColor,
            averageCycleLength = averageCycleLength,
            averagePeriodLength = averagePeriodLength,
            lutealPhaseLength = lutealPhaseLength,
            temperatureUnit = temperatureUnit.name,
            dateFormat = dateFormat,
            language = language,
            isPinEnabled = isPinEnabled,
            pinHash = pinHash,
            isBiometricEnabled = isBiometricEnabled,
            isPrivacyModeEnabled = isPrivacyModeEnabled,
            hideNotificationContent = hideNotificationContent,
            notificationsEnabled = notificationsEnabled,
            quietHoursEnabled = quietHoursEnabled,
            quietHoursStart = quietHoursStart.toString(),
            quietHoursEnd = quietHoursEnd.toString(),
            defaultReminderTime = defaultReminderTime.toString(),
            medicationReminderTime = medicationReminderTime.toString(),
            hydrationReminderTime = hydrationReminderTime.toString(),
            bodyMetricsReminderTime = bodyMetricsReminderTime.toString(),
            onboardingCompleted = onboardingCompleted,
            profileName = userProfile.name,
            profileBirthYear = userProfile.birthYear,
            profileTryingToConceive = userProfile.tryingToConceive,
            profileTimezoneId = userProfile.timezoneId,
            pregnancyMode = pregnancyMode,
            breastfeedingMode = breastfeedingMode,
            menopauseMode = menopauseMode
        )
    }
}

File: cyclecare_flutter/lib/data/repositories/settings_repository_impl.dart
import '../../domain/entities/settings.dart' as domain;
import '../../domain/repositories/settings_repository.dart';
import '../database/app_database.dart';

class SettingsRepositoryImpl implements SettingsRepository {
  final AppDatabase _database;

  SettingsRepositoryImpl(this._database);

  @override
  Future<domain.Settings?> getSettings() async {
    final settings = await _database.getSettings();
    return settings != null ? _toDomain(settings) : null;
  }

  @override
  Stream<domain.Settings?> watchSettings() {
    return _database.watchSettings().map(
          (settings) => settings != null ? _toDomain(settings) : null,
        );
  }

  @override
  Future<bool> updateSettings(domain.Settings settings) {
    return _database.updateSettings(_toData(settings));
  }

  @override
  Future<void> updateCycleSettings(int cycleLength, int periodLength) async {
    final current = await getSettings();
    if (current != null) {
      await updateSettings(
        current.copyWith(
          averageCycleLength: cycleLength,
          averagePeriodLength: periodLength,
        ),
      );
    }
  }

  @override
  Future<void> updatePrivacySettings(
      bool pinEnabled, bool biometricEnabled, bool privacyMode) async {
    final current = await getSettings();
    if (current != null) {
      await updateSettings(
        current.copyWith(
          isPinEnabled: pinEnabled,
          isBiometricEnabled: biometricEnabled,
          isPrivacyModeEnabled: privacyMode,
        ),
      );
    }
  }

  @override
  Future<void> updateNotificationSettings(
      bool enabled, bool quietHours, String start, String end) async {
    final current = await getSettings();
    if (current != null) {
      await updateSettings(
        current.copyWith(
          notificationsEnabled: enabled,
          quietHoursEnabled: quietHours,
          quietHoursStart: start,
          quietHoursEnd: end,
        ),
      );
    }
  }

  @override
  Future<void> completeOnboarding() async {
    final current = await getSettings();
    if (current != null) {
      await updateSettings(
        current.copyWith(onboardingCompleted: true),
      );
    }
  }

  domain.Settings _toDomain(Setting setting) {
    return domain.Settings(
      id: setting.id,
      theme: setting.theme,
      primaryColor: setting.primaryColor,
      averageCycleLength: setting.averageCycleLength,
      averagePeriodLength: setting.averagePeriodLength,
      lutealPhaseLength: setting.lutealPhaseLength,
      temperatureUnit: setting.temperatureUnit,
      dateFormat: setting.dateFormat,
      language: setting.language,
      isPinEnabled: setting.isPinEnabled,
      pinHash: setting.pinHash,
      isBiometricEnabled: setting.isBiometricEnabled,
      isPrivacyModeEnabled: setting.isPrivacyModeEnabled,
      hideNotificationContent: setting.hideNotificationContent,
      notificationsEnabled: setting.notificationsEnabled,
      quietHoursEnabled: setting.quietHoursEnabled,
      quietHoursStart: setting.quietHoursStart,
      quietHoursEnd: setting.quietHoursEnd,
      onboardingCompleted: setting.onboardingCompleted,
      profileName: setting.profileName,
      profileBirthYear: setting.profileBirthYear,
      profileTryingToConceive: setting.profileTryingToConceive,
      pregnancyMode: setting.pregnancyMode,
      breastfeedingMode: setting.breastfeedingMode,
      menopauseMode: setting.menopauseMode,
    );
  }

  Setting _toData(domain.Settings settings) {
    return Setting(
      id: settings.id,
      theme: settings.theme,
      primaryColor: settings.primaryColor,
      averageCycleLength: settings.averageCycleLength,
      averagePeriodLength: settings.averagePeriodLength,
      lutealPhaseLength: settings.lutealPhaseLength,
      temperatureUnit: settings.temperatureUnit,
      dateFormat: settings.dateFormat,
      language: settings.language,
      isPinEnabled: settings.isPinEnabled,
      pinHash: settings.pinHash,
      isBiometricEnabled: settings.isBiometricEnabled,
      isPrivacyModeEnabled: settings.isPrivacyModeEnabled,
      hideNotificationContent: settings.hideNotificationContent,
      notificationsEnabled: settings.notificationsEnabled,
      quietHoursEnabled: settings.quietHoursEnabled,
      quietHoursStart: settings.quietHoursStart,
      quietHoursEnd: settings.quietHoursEnd,
      onboardingCompleted: settings.onboardingCompleted,
      profileName: settings.profileName,
      profileBirthYear: settings.profileBirthYear,
      profileTryingToConceive: settings.profileTryingToConceive,
      pregnancyMode: settings.pregnancyMode,
      breastfeedingMode: settings.breastfeedingMode,
      menopauseMode: settings.menopauseMode,
    );
  }
}
